src/client_side.cc

```diff
@@ -3478,6 +3482,15 @@ clientListenerConnectionOpened(AnyP::PortCfgPointer &s, const Ipc::FdNoteId port
            << s->listenConn);
 
     Must(AddOpenedHttpSocket(s->listenConn)); // otherwise, we have received a fd we did not ask for
+
+#if USE_SYSTEMD
+    // at least one port is open, tell systemd we are able to serve connections
+    // subsequent sd_notify() calls will not be processed because first parameter is non-zero
+    // XXX: send the notification only when all configured ports are open
+    static int portsSeen = 0;
+    if (++portsSeen == NHttpSockets && (opt_foreground || opt_no_daemon))
```

Comment:
The `++portsSeen == NHttpSockets` condition is very complex. It is unlikely that somebody will understand what it really means or why it is here without spending significant time on researching the answer. Can we simply remove it?

```suggestion
    // When the very first port opens, tell systemd we are able to serve connections.
    // Subsequent sd_notify() calls, including calls during reconfiguration,
    // do nothing because the first call parameter is 1.
    // XXX: Send the notification only after opening all configured ports.
    if (opt_foreground || opt_no_daemon)
```

Reply:
It was Amos suggestion. I can remove it if you both agree.

Reply:
> I can remove it if you both agree.

Sounds good -- you are in charge of making that agreement happen. My current arguments are reflected in the change request. FWIW, I will withdraw my change request if it is wrong, of course.

Reply:
Update: I found two more problems with the `++portsSeen == NHttpSockets` condition:

First of all, Squid does not call sd_notify() at all if the last of the configured ports fails to open (but the first one(s) have opened). This was one of the problems I warned about in my [item 2](https://bugs.squid-cache.org/show_bug.cgi?id=5016#c6)  -- correctly tracking when the last port that could open was opened "requires a bit of development to track the number of ports that failed to open". The simpler first item on that list is what would happen if you remove the portsSeen condition.

```
2020/01/13 14:57:36.188 kid1| Accepting HTTP Socket connections at conn3 local=[::]:3131 remote=[::] FD 29 flags=1
2020/01/13 14:57:36.188 kid1| XXX: opt_foreground=1 opt_no_daemon=0 portsSeen=1 NHttpSockets=2
...
2020/01/13 14:57:36.191 kid4| commBind Cannot bind socket FD 14 to [::]:22: (98) Address already in use
```

I added the `XXX` debugging before the sd_notify() call to illustrate that portsSeen does not reach NHttpSockets=2 when the last port fails to open. In this specific test, two listening ports were configured, and the second port (in the squid.conf order) was already occupied at Squid start time. Squid pretends to run fine in this case, so systemd has to be notified (my footnote [1](https://bugs.squid-cache.org/show_bug.cgi?id=5016#c6) talks more about that pretending).

Second, the condition does not match its `at least one port is open, tell systemd` documentation. The condition only becomes true only if _all_ configured ports opened successfully, not if at least one port was opened.

Reply:
Thank you for fixing this in d5f4a31. I am resolving my change request.